theforeman.org/pipelines/lib/stable/smart-proxy.groovy

```diff
@@ -0,0 +1,3 @@
+def git_checkout() {
+    git(url: 'https://github.com/theforeman/smart-proxy', branch: git_branch)
```

Comment:
@evgeni I can't quite figure out how I would get `git_branch` from the top level into the function. Adding a parameter like this doesn't work either. Any suggestion?
```suggestion
def git_checkout(git_branch=git_branch) {
    git(url: 'https://github.com/theforeman/smart-proxy', branch: git_branch)
```

Reply:
I'll try to think tomorrow ;)

Reply:
I think the only way to do it is to define `git_checkout` with a `git_branch` parameter and then actually pass that from the caller:


```groovy
def git_checkout(git_branch) {
…
}
…
                stage('Setup Git Repos') {
                    steps {
                        git_checkout(git_branch)
                    }
                }
```

Reply:
The problem with that is that `git_branch` is undefined for PR builds.

https://stackoverflow.com/questions/6305910/how-do-i-create-and-access-the-global-variables-in-groovy suggests that if you use `git_branch = '..'` instead of `def git_branch = '..'` it may work. I'll try that first.

Reply:
That works so I think this PR is now ready.

Reply:
Relying on a "global" to exist is generally a bad pattern and can lead to a variety of issues from my experience. Reading back, I interpreted this as -- sometimes git_branch will be defined and sometimes it will be effectively `null` ?

Reply:
> The problem with that is that `git_branch` is undefined for PR builds.

Why is this true? I would have thought in a PR test the git_branch would be defined as the either the target branch or the branch of the PR itself from the users fork.

Reply:
What about something like:

```
def git_checkout(git_branch) {
    if (git_branch) {
           git(url: 'https://github.com/theforeman/smart-proxy', branch: git_branch)
    } else {
           ghprb_git_checkout()
    }
}
```

You could even make debugging a bit easier:

```
def git_checkout(git_branch) {
    if (git_branch) {
           git(url: 'https://github.com/theforeman/smart-proxy', branch: git_branch)
    } else if (ghprbTargetBranch) {
           ghprb_git_checkout()
    } else {
          error("No git_branch or PR target branch specified")
    }
}
```

Reply:
I suppose I could use `git_branch = env.ghprbTargetBranch` or `git_branch = null`, but it feels wrong if it's not going to be used anyway. You would end up with:

```groovy
def git_checkout(git_branch) = {
    ghprb_git_checkout()
}
```
I also thought about [currying](https://groovy-lang.org/closures.html#_currying) but you can't do that outside of script blocks (since you can't invoke function).

I'm beginning to think we need libraries. However, I just found https://plugins.jenkins.io/git-parameter/ which we have installed, so I'm going to play with that as well.

Reply:
How about this version?